pkg/bigquery/sanitize.go

```diff
@@ -0,0 +1,99 @@
+// Copyright 2023 The Authors (see AUTHORS file)
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package bigquery implements helpers for using BigQuery.
+package bigquery
+
+import (
+	"fmt"
+	"regexp"
+	"unicode/utf8"
+)
+
+const tableNameMaxUTF8Bytes = 1024
+
+// Start with lowercase, middle is lowercase, number or hyphen, cannot end in
+// hyphen. 6-30 characters in length (start, 4-28 middle, end).
+var projectIDMatcher = regexp.MustCompile(`^[a-z][a-z0-9\-]{4,28}[a-z0-9]$`)
+
+// Lowercase and uppercase letters and underscores. Max 1024 characters.
+// regexp only allows 1000 repetitions, so had to manually repeat.
+const datasetIDRegex = `^[a-zA-Z0-9_]{1,512}[a-zA-Z0-9_]{0,512}$`
+
+var datasetIDMatcher = regexp.MustCompile(datasetIDRegex)
+
+// Unicode characters in category L (letter), M (mark), N (number),
+// Pc (connector, including underscore), Pd (dash), Zs (space). Max 1024
+// characters. 1024 is actually UTF-8 bytes from experimentation.
+// UTF-8 length check will be done separately.
+// regexp only allows 1000 repetitions, so had to manually repeat.
+const tableNameRegex = `^[\p{L}\p{M}\p{N}\p{Pc}\p{Pd}\p{Zs}]{1,512}[\p{L}\p{M}\p{N}\p{Pc}\p{Pd}\p{Zs}]{0,512}$`
```

Comment:
Since the length is already enforced elsewhere, it might be unnecessary and potentially confusing to also enforce it here. Especially with the hack to work around the limitation of 1000 repetitions.

Reply:
That's a good point. I'll change it.